micro-rdk/src/esp32/ota.rs

```diff
@@ -0,0 +1,189 @@
+        let kind = ota_config
+            .attributes
+            .as_ref()
+            .unwrap()
+            .fields
+            .get("url")
+            .unwrap()
+            .kind
+            .clone()
+            .unwrap();
+        let url = match kind {
+            crate::google::protobuf::value::Kind::StringValue(s) => s,
+            _ => "".to_string(),
+        };
+        let uri = url.parse::<hyper::Uri>().unwrap();
+
+        Self {
+            url: uri.to_string(),
+        }
+    }
+
+    pub(crate) async fn update(&mut self) -> Result<(), OtaError> {
+        let connection = EspHttpConnection::new(&Configuration {
+            buffer_size: Some(OTA_HTTP_BUFSIZE),
+            buffer_size_tx: Some(OTA_HTTP_BUFSIZE),
+            use_global_ca_store: true,
+            crt_bundle_attach: Some(esp_idf_svc::sys::esp_crt_bundle_attach),
+            ..Default::default()
+        })
+        .map_err(|e| OtaError::DownloadError(e.to_string()))?;
+
+        let mut client = Client::wrap(connection);
+
+        let headers = [("accept", "application/octet_stream")];
+        let request = client
+            .request(Method::Get, &self.url, &headers)
+            .map_err(|e| OtaError::DownloadError(e.to_string()))?;
+
+        let mut response = request
+            .submit()
+            .map_err(|e| OtaError::DownloadError(e.to_string()))?;
+        let status = response.status();
+
+        if status < 200 || status > 299 {
+            return Err(OtaError::DownloadError(
+                format!("Bad Request - Status:{}", status).to_string(),
+            ));
+        }
+
+        let file_len = response.content_len().unwrap_or(0) as usize;
+
+        if file_len <= OTA_MIN_SIZE {
+            log::error!("new image too small: {} bytes", file_len);
+            return Err(OtaError::InvalidImage);
+        }
+        if file_len > OTA_MAX_SIZE {
+            log::error!("new image too big: {} bytes", file_len);
+            return Err(OtaError::InvalidImage);
+        }
+
+        let mut ota = EspOta::new().map_err(OtaError::EspError)?;
+        let running_fw_info = ota.get_running_slot().map_err(OtaError::EspError)?.firmware;
+        let mut update_handle = ota.initiate_update().map_err(|_| OtaError::InitError)?;
+        let mut buff = vec![0; OTA_CHUNK_SIZE];
+        let mut total_read_len: usize = 0;
+        let mut got_info = false;
+        while total_read_len < file_len {
+            let n = response.read(&mut buff).unwrap_or_default();
+            total_read_len += n;
+            if !got_info {
+                let mut loader = EspFirmwareInfoLoader::new();
+                loader.load(&mut buff).map_err(OtaError::EspError)?;
+                let new_fw = loader.get_info().map_err(OtaError::EspError)?;
+
+                log::info!("Firmware to be downloaded: {new_fw:?}");
+                if let Some(ref running_fw) = running_fw_info {
+                    if running_fw.version == new_fw.version {
```

Comment:
should we compare `signature` instead of `version`?

Reply:
it's all 0's for some reason. still unsure of how this field is populated. The `version` field is currently a combination of build-time info and timestamp.